File: Assignment6.rs
use rand::Rng; //random
use std::time::Instant; //time measurement

fn fill(arr: &mut [i32]) { //fill array with mutible ints with 32 bits
    let mut rng = rand::thread_rng();
    for i in arr.iter_mut() {
        *i = rng.gen_range(1..=1000);
    }
}

fn binary_search(arr: &[i32], target: i32) -> Option<usize> { //binary search function
    let mut left = 0;
    let mut right = arr.len() - 1;

    while left <= right {
        let mid = left + (right - left) / 2;

        if arr[mid] == target {
            return Some(mid);
        } else if arr[mid] < target {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    None
}

fn main() {
    let arrays = vec![ //mutable arrays
        vec![0; 100],
        vec![0; 200],
        vec![0; 400],
        vec![0; 800],
        vec![0; 1600],
        vec![0; 3200],
        vec![0; 6400],
        vec![0; 12800],
        vec![0; 25600],
        vec![0; 51200],
    ];

    let target = 2_147_483_647; //impossible number to find

    let mut durations = Vec::new(); //time measurement

    for arr in &arrays { //main
        fill(&mut arr.clone());
        let start = Instant::now();
        for _ in 0..10_000_000 {
            let _ = binary_search(&arr, target);
        }
        let end = Instant::now();
        let duration = end.duration_since(start).as_nanos();
        durations.push(duration);
    }

    for (i, duration) in durations.iter().enumerate() {
        println!("Time {}: {} nanoseconds", i + 1, duration);
    }
}

/*
DATA STRUCTURES USED - Vectors, Arrays

The provided code performs binary search on different mutable arrays of varying sizes and 
measures the execution time for each search operation. It uses the fill function to populate 
the arrays with random integers and the binary_search function to perform the binary 
search algorithm. The arrays are defined as a vector of mutable vectors (arrays). The 
code iterates over each array, fills it with random values, performs 10,000,000 search 
operations on each array, and records the duration of each search. The durations are stored 
in a separate vector (durations) and then displayed on the console.

Time Complexity:
The time complexity of the binary search algorithm is O(log n), where n is the size of the 
input array. In this code, binary search is performed 10,000,000 times for each array in 
the arrays vector. Therefore, the time complexity of the code can be considered as 
O(10,000,000 * log n), where n represents the size of each array.

Space Complexity:
The space complexity of the code is determined by the space used to store the arrays and the 
durations vector. The space required for the arrays is proportional to the total number of 
elements across all arrays. The space required for the durations vector is proportional to 
the number of arrays. Therefore, the space complexity can be considered as O(N), where N 
is the total number of elements across all arrays plus the number of arrays.

Run 1::
Time 1: 1.1925203 seconds
Time 2: 1.4850099 seconds
Time 3: 1.5953257 seconds
Time 4: 1.7131715 seconds
Time 5: 2.0921714 seconds
Time 6: 2.8639674 seconds
Time 7: 2.437687 seconds
Time 8: 2.4179654 seconds
Time 9: 2.577816 seconds
Time 10: 2.7117438 seconds

Run2::
Time 1: 1.6071248 seconds
Time 2: 1.4526246 seconds
Time 3: 1.7735745 seconds
Time 4: 1.9741895 seconds
Time 5: 1.9435543 seconds
Time 6: 2.2024632 seconds
Time 7: 2.1590178 seconds
Time 8: 2.1225288 seconds
Time 9: 2.5231159 seconds
Time 10: 2.7715585 seconds

Run3::
Time 1: 1.5025101 seconds
Time 2: 1.5018165 seconds
Time 3: 2.18645 seconds
Time 4: 2.513827 seconds
Time 5: 2.2210145 seconds
Time 6: 2.2992594 seconds
Time 7: 2.240435 seconds
Time 8: 2.5571263 seconds
Time 9: 2.4283183 seconds
Time 10: 2.4862328 seconds
*/

File: Assignment6.py
import random
import time

def fill(arr): #initialize arrays
    for i in range(len(arr)):
        arr[i] = random.randint(1, 1000)

def binary_search(arr, target): #binary search function
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = left + (right - left) // 2

        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1

arrays = [[100], [200], [400], [800], [1600], [3200],
          [6400], [12800], [25600], [51200]] #all arrays doubling until 51200

target = 2147483647 #target that we won't find

durations = []

for arr in arrays: #main
    fill(arr)
    start = time.perf_counter()
    for _ in range(10000000):
        index = binary_search(arr, target)
    end = time.perf_counter()
    durations.append((end - start))  # In seconds

for i, duration in enumerate(durations):
    print(f"Time {i + 1}: {duration} seconds")

"""
DATA STRUCTURES USED - Lists

The given code performs binary search on different arrays of varying sizes and measures 
the execution time for each search operation. It uses the fill function to initialize 
the arrays with random values and the binary_search function to perform the binary search 
algorithm. The arrays are defined as a list of lists (arrays), where each inner list 
represents an array with a specific size. The code iterates over each array, performs 
10,000,000 search operations for each array, and records the duration of each search. 
The durations are stored in a separate list (durations) and then displayed on the console.

Time Complexity:
The time complexity of the binary search algorithm is O(log n), where n is the size of the 
input array. In this code, binary search is performed 10,000,000 times for each array in the 
arrays list. Therefore, the time complexity of the code is O(10,000,000 * log n), where n 
represents the size of each array.

Space Complexity:
The space complexity of the code is determined by the space used to store the arrays and 
the durations list. The space required for the arrays is proportional to the total number 
of elements across all arrays. The space required for the durations list is proportional 
to the number of arrays. Therefore, the space complexity can be considered as O(N), where 
N is the total number of elements across all arrays plus the number of arrays.

Run 1::
Time 1: 4.6877584 seconds
Time 2: 4.8131024 seconds
Time 3: 4.7948068 seconds
Time 4: 4.866584999999999 seconds
Time 5: 4.7196494000000015 seconds
Time 6: 4.6712983 seconds
Time 7: 4.765804000000003 seconds
Time 8: 4.740078799999999 seconds
Time 9: 4.805248899999995 seconds
Time 10: 4.699169300000001 seconds

Run2::
Time 1: 4.7471711 seconds
Time 2: 4.8277282999999995 seconds
Time 3: 4.757702399999999 seconds
Time 4: 4.861356799999999 seconds
Time 5: 4.986174000000002 seconds
Time 6: 4.759046100000003 seconds
Time 7: 4.805968200000002 seconds
Time 8: 4.853549100000002 seconds
Time 9: 4.789194599999995 seconds
Time 10: 4.840418200000002 seconds

Run3::
Time 1: 4.8334048 seconds
Time 2: 4.7744188 seconds
Time 3: 4.8920178 seconds
Time 4: 4.7807466000000005 seconds
Time 5: 4.808437699999999 seconds
Time 6: 5.002357700000001 seconds
Time 7: 4.862001900000003 seconds
Time 8: 4.741033300000005 seconds
Time 9: 4.766355500000003 seconds
Time 10: 4.8410054 seconds
"""